examples/tidalfarm/tidalfarm.py

```diff
@@ -1,119 +1,189 @@
+solver_obj.eq_sw.add_term(turbine_drag_term, 'implicit')
+
+turbine_friction.assign(0.0)
+solver_obj.assign_initial_conditions(uv=as_vector((1e-7, 0.0)))
+
+# Setup the functional. It computes a measure of the profit as the difference
+# of the power output of the farm (the "revenue") minus the cost based on the number
+# of turbines
+
+u, eta = split(solver_obj.fields.solution_2d)
+# should multiply this by density to get power in W - assuming rho=1000 we get kW instead
+power_integral = turbine_friction * (u[0]*u[0] + u[1]*u[1])**1.5 * dx(2)
+
+# turbine friction=C_T*A_T/2.*turbine_density
+C_T = 0.8  # turbine thrust coefficient
+A_T = pi * (16./2)**2  # turbine cross section
+# cost integral is n/o turbines = \int turbine_density = \int c_t/(C_T A_T/2.)
+cost_integral = 1./(C_T*A_T/2.) * turbine_friction * dx(2)
 
-def jfunc(m):
-    drag_func.assign(m)
-    solver_obj.simulation_time = 0.
-    solver_obj.iteration = 0
-    solver_obj.i_export = 0
-    solver_obj.assign_initial_conditions(uv=as_vector((velocity_u, 0.0)), elev=Constant(0.0))
-    solver_obj.iterate()
-    Jm = assemble(integral)
-    return Jm
+break_even_wattage = 10  # (kW) amount of power produced per turbine on average to "break even" (cost = revenue)
 
+# we rescale the functional such that the gradients are ~ order magnitude 1.
+# the scaling is chosen such that the gradient of break_even_wattage * cost_integral is of order 1
+# the power-integral is assumed to be of the same order of magnitude
+scaling = 1./assemble(break_even_wattage/(C_T*A_T/2.) * dx(2, domain=mesh2d))
+combined_functional = Functional(scaling * (power_integral - break_even_wattage * cost_integral) * dt)
 
-success = replay_dolfin(tol=0.0, stop=False)
-print_output(solver_obj.fields.solution_2d.vector().array()[0:10])
-Jhat = ReducedFunctional(J, c)
-print_output("Output of Jhat: ", Jhat(drag_func))
-minconv = taylor_test(jfunc, c, J0, dJdc, seed=1e-4)
-assert minconv > 1.95
+
+# a function to update the tidal_elev bc value every timestep
+# we also use it to display the profit each time step (which will be a time-integrated into the functional)
+x = SpatialCoordinate(mesh2d)
+g = 9.81
+omega = 2 * pi / tidal_period
+
+
+def update_forcings(t, annotate=True):
+    print "Updating tidal elevation at t = ", t
+    P = assemble(power_integral)
+    N = assemble(cost_integral)
+    print "Power, N turbines, profit =", P, N, P-break_even_wattage*N
+    # NOTE: we need to explicitly tell dolfin-adjoint to annotate this as by default it seems to
+    # only annotate if the interpoland is a Function (is this reasonable?)
+    tidal_elev.interpolate(tidal_amplitude*sin(omega*t + omega/pow(g*H, 0.5)*x[0]), annotate=annotate)
+    # NOTE: it seems firedrake-adjoint (dolfin-adjoint in general?) cannot handle functions that are part of a
+    # time-integrated Functional which do not change over time and are therefore not annotated
+    turbine_friction.project(turbine_friction, annotate=annotate)
```

Comment:
I'll give that a try. I'm not entirely sure it's going to work though: the project call here is just a work-around for a bug in dolfin-adjoint: if turbine_friction is not touched during a timestep, dolfin-adjoint fails to do the time-integral as it gets confused about what the value of turbine_friction during each timestep should be. I previously worked around this with a similar interpolate call which fixed the error message, but gave the wrong answer! I will submit a bug-report to dolfin-adjoint as well.